In [58]:
import pandas as pd
import kagglehub
from sklearn.datasets import fetch_california_housing

# Load the housing dataset
housing = fetch_california_housing()
print(housing.DESCR)


.. _california_housing_dataset:

California Housing dataset
--------------------------

**Data Set Characteristics:**

:Number of Instances: 20640

:Number of Attributes: 8 numeric, predictive attributes and the target

:Attribute Information:
    - MedInc        median income in block group
    - HouseAge      median house age in block group
    - AveRooms      average number of rooms per household
    - AveBedrms     average number of bedrooms per household
    - Population    block group population
    - AveOccup      average number of household members
    - Latitude      block group latitude
    - Longitude     block group longitude

:Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

The target variable is the median house value for California districts,
expressed in hundreds of thousands of dollars ($100,000).

This dataset was derived from the 1990 U.S. census, using one row per ce

In [ ]:
# Organize the data into a data frame and series for X and y variables
X = pd.DataFrame(housing.data, columns=housing.feature_names) 
y = pd.Series(housing.target, name='med_house_value')

# Display the first few rows of the feature and target data
print("\nFirst 5 rows of the feature data:")
print(X.head())

# Print the feature names in the dataset
print("Feature names in the California housing dataset:")
print(housing.feature_names)

#Print the first 5 rows of the target data
print("\nFirst 5 rows of the target data (what we are trying to predict):")
print(y.head())



First 5 rows of the feature data:
   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25  
Feature names in the California housing dataset:
['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']

First 5 rows of the target data (what we are trying to predict):
0    4.526
1    3.585
2    3.521
3    3.413
4    3.422
Name: med_house_value, dtype: float64


In [35]:
#Check for missing values
print("Missing values in the feature data:")
print(X.isnull().sum())
print("\nMissing values in the target data:")
print(y.isnull().sum())


Missing values in the feature data:
MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
dtype: int64

Missing values in the target data:
0


In [40]:
#Print summary statistics of the feature data
print("\nSummary statistics of the feature data:")
print(X.describe())
#Print summary statistics of the target data
print("\nSummary statistics of the target data:")
print(y.describe())


Summary statistics of the feature data:
             MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude  
count  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704  
std       10.386050      2.135952      2.003532  
min        0.692308     32.540000   -124.350000

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# Split ther aw data into a training (80%) and testing (20%) set 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
trained_set = LinearRegression()
trained_set.fit(X_train, y_train)

#use the regression model to make predictions

# Make predictions on the test set
y_pred = trained_set.predict(X_test)


[0.71912284 1.76401657 2.70965883 ... 4.46877017 1.18751119 2.00940251]


In [ ]:
#evaluate the model's performance
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score

mse = mean_squared_error(y_test, y_pred)
rmse = root_mean_squared_error(y_test, y_pred)
r2_raw = r2_score(y_test, y_pred)

print("Unscaled Data Model:")
print(f"Mean Squared Error: {mse:.2f}")
print(f"Root Mean Squared Error: {rmse:.2f}")
print(f"R² Score: {r2_raw:.2f}")

#Veiw our coefficients
print("Model Coefficients (Unscaled):")
print(pd.Series(trained_set.coef_, index=X.columns))




Unscaled Data Model:
Mean Squared Error: 0.56
Root Squared Error: 0.75
R² Score: 0.58
Model Coefficients (Unscaled):
MedInc        0.448675
HouseAge      0.009724
AveRooms     -0.123323
AveBedrms     0.783145
Population   -0.000002
AveOccup     -0.003526
Latitude     -0.419792
Longitude    -0.433708
dtype: float64


Interpertation:
 When looking at the MSE, each variables coefficient represents the change in the (target variable) medium house value, for a one-unit change in the feature measured. When observing our data we see that a increase in the average number of bedrooms per houshold, results in the median house value rising by $78,000 in the census block. Additionally a 100,000 increase in the median income for a block (a 1 unit increase) results in the growth of ~$44,000 in median house value assuming other factors remain constant. It is important to note that because some of these features are in different units of measure, we may be mislead if we compare each cofficent.

The R^2 value of 0.58 tells us that the model explains about 58% of the variance in the house values based on the features we have used. The closer to 1 the R^2 the better the fit of the regession line so figure of 0.58 suggests that there is still a significant portion of the variance that remains unexplained by our model.

The average bedroom per houshold as well as the median income per household seem to display the greatest postive impact on housing values at 
0.78 and 0.44 respectivley. Additionally, Latitude and Logitude by coefficient's of -0.419 and -0.433, displaying a negative impact on home values as the directions increase. The lowest impact on our regression is in population and house age, indicating these feature may not have a strong impact on the predictive power of the model.

The last thing I will note is the MRSE is 0.75 which tells us that, on average, each plot point is off the regression line by ~75k. This seems to be a signifcant weakness of the model, especially because the range of the home prices are between 150k and 500k

In [69]:
#Select three features from the dataset to build a simplified model.
feature_vars = ["MedInc", "AveRooms", "Longitude"]

X_new = pd.DataFrame(housing.data, columns=housing.feature_names)[feature_vars]
y_new = pd.Series(housing.target, name='med_house_value')
print("Dataset preview:")
print(X_new.head())

#Split the new data into training and testing sets
X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(X_new, y_new, test_size=0.2, random_state=42)

# Make predictions on the test set
new_trained_set = LinearRegression()
new_trained_set.fit(X_train_new, y_train_new)

y_pred_new = new_trained_set.predict(X_test_new)
mse_new = mean_squared_error(y_test_new, y_pred_new)
rmse_new = root_mean_squared_error(y_test_new, y_pred_new)
r2_new = r2_score(y_test_new, y_pred_new)

print("\nUnscaled Data Model:")
print(f"Mean Squared Error: {mse_new:.2f}")
print(f"Root Mean Squared Error: {rmse_new:.2f}")
print(f"R² Score: {r2_new:.2f}")




Dataset preview:
   MedInc  AveRooms  Longitude
0  8.3252  6.984127    -122.23
1  8.3014  6.238137    -122.22
2  7.2574  8.288136    -122.24
3  5.6431  5.817352    -122.25
4  3.8462  6.281853    -122.25

Unscaled Data Model:
Mean Squared Error: 0.70
Root Mean Squared Error: 0.84
R² Score: 0.47


How does the simplified model compare to the full model?
The simplifed model has less explanatory power then the model with 8 features. This is displayed through a lower R^2 and a higher RMSE and MSE. When only using three features the % of variance explained by our model decreases from 58% to 47% and on average, each plot point is further in the simplified model from the regession line then it was in the full model.

Would you use this simplified model in practice? Why or why not?
I would not choose to use the simple model. It has lower test statistics indicating that more factor are benefical when makeing prediction. In the model with more factors the evaluation statistics are higher, indicatting a better fit with the data.